runtime/common/src/slots.rs

```diff
@@ -427,20 +431,34 @@ impl<T: Config> Leaser for Pallet<T> {
 			.unwrap_or_else(Zero::zero)
 	}
 
-	fn lease_period() -> Self::LeasePeriod {
-		T::LeasePeriod::get()
+	#[cfg(any(feature = "runtime-benchmarks", test))]
+	fn lease_period_length() -> (T::BlockNumber, T::BlockNumber) {
+		(T::LeasePeriod::get(), T::LeaseOffset::get())
 	}
 
-	fn lease_period_index() -> Self::LeasePeriod {
-		<frame_system::Pallet<T>>::block_number() / T::LeasePeriod::get()
+	fn lease_period_index(b: T::BlockNumber) -> (Self::LeasePeriod, bool) {
+		// Note that lease lease period 0 is artificially extended by the lease offset.
+		let offset_block_now = b.saturating_sub(T::LeaseOffset::get());
+		let lease_period = offset_block_now / T::LeasePeriod::get();
+
+		// Special logic to handle lease period 0 being extended by the offset.
+		let first_block = if lease_period.is_zero() {
+			// Lease period 0 has their first block on only block 0.
+			b.is_zero()
+		} else {
+			(offset_block_now % T::LeasePeriod::get()).is_zero()
+		};
+
+		(lease_period, first_block)
```

Comment:
Here is the core change. This is the one source of truth for calculating the current lease period index given some offset. All other APIs must use this to do stuff.